nixos/modules/services/web-servers/tomcat/default.nix

```diff
@@ -0,0 +1,480 @@
+      };
+    in
+      ''
+        <SSLHostConfig ${concatStringsSep " " (mapAttrsToList (k: v: "${k}=\"${toStr v}\"") (defaultAttrs // hostOpts.sslHostConfig))}>
+          <Certificate
+            certificateKeyFile="${sslServerKey}"
+            certificateFile="${sslServerCert}"
+            certificateChainFile="${sslServerChain}"
+          />
+        </SSLHostConfig>
+      ''
+  ;
+
+  mkVHostConf = hostOpts:
+    let
+      useACME = hostOpts.enableACME; # TODO: don't try to serve up directories that don't exist - ie. are we using DNS challenge? || hostOpts.useACMEHost != null;
+      acmeChallenge = optionalString useACME ''
+        <Context docBase="${hostOpts.acmeRoot}/.well-known/acme-challenge" path="/.well-known/acme-challenge" />
+      '';
+      robotsEntries = optionalString (hostOpts.robotsEntries != "") ''
+        <Context docBase="${pkgs.writeText "robots.txt" hostOpts.robotsEntries}" path="/robots.txt" />
+      '';
+
+      # workDir and xmlBase allow tomcat to function with an immutable configuration directory
+      defaultAttrs = {
+        name = hostOpts.hostName;
+        appBase = hostOpts.appBase;
+        unpackWARs = true;
+        autoDeploy = true;
+        xmlBase = "${stateDir}/CATALINA/${hostOpts.hostName}";
+        workDir = "${cacheDir}/${hostOpts.hostName}";
+      };
+    in
+      ''
+        <Host ${concatStringsSep " " (mapAttrsToList (k: v: "${k}=\"${toStr v}\"") (defaultAttrs // hostOpts.settings))}>
+          ${concatMapStrings (alias: "<Alias>${alias}</Alias>") hostOpts.serverAliases}
+          ${acmeChallenge}
+          ${robotsEntries}
+          <Valve
+            className="org.apache.catalina.valves.AccessLogValve"
+            directory="${logsDir}"
+            prefix="${hostOpts.hostName}_access" suffix=".log"
+            pattern="%h %l %u %t &quot;%r&quot; %s %b"
+          />
+          ${hostOpts.extraConfig}
+        </Host>
+      ''
+  ;
+
+  serverXmlFile = pkgs.writeText "server.xml" ''
+    <?xml version="1.0" encoding="UTF-8"?>
+    <!--
+      This is a generated file.  Do not edit!
+
+      To make changes, edit the services.tomcat NixOS options
+      in your /etc/nixos/configuration.nix file.
+    -->
+    <Server ${concatStringsSep " " (mapAttrsToList (k: v: "${k}=\"${toStr v}\"") cfg.serverAttrs)}>
+      ${cfg.extraConfig}
+
+      <Service name="Catalina">
+        ${concatMapStrings mkConnectorConf cfg.connectors}
+
+        <Engine ${concatStringsSep " " (mapAttrsToList (k: v: "${k}=\"${toStr v}\"") cfg.engineAttrs)}>
+          ${concatMapStrings mkVHostConf vhosts}
+        </Engine>
+      </Service>
+    </Server>
+  '';
+
+  # TODO: write a custom logging.properties file i guess
+  # https://github.com/GoogleCloudPlatform/tomcat-runtime/pull/28/files
+  loggingProperties = pkgs.writeText "logging.properties" ''
+    handlers = java.util.logging.ConsoleHandler
```

Comment:
yes, would be nice for this to be configurable. I had this replaced by
```
             # this configures such tomcat log files:
             # - catalina.log: contains all tomcat messages
             # - catalina.out: all System.out and System.err messages
             # catalina.log and catalina.out have no rotation, so logrotate must be used to limit files by size
             # (rumors on internet say tomcat crashes when log files are > 2G)

             handlers = 1catalina.org.apache.juli.AsyncFileHandler
             .handlers = 1catalina.org.apache.juli.AsyncFileHandler

             1catalina.org.apache.juli.AsyncFileHandler.level = FINE
             1catalina.org.apache.juli.AsyncFileHandler.directory = $\{catalina.base}/logs
             1catalina.org.apache.juli.AsyncFileHandler.rotatable = false
             1catalina.org.apache.juli.AsyncFileHandler.prefix = catalina
```

Reply:
Now fully configurable, but defaults to logging everything except access logs to `journald` :+1: